224-basic-calculator.py

```diff
@@ -0,0 +1,61 @@
+# 224. Basic Calculator
+
+class Solution:
+    def calculate(self,s):
+        """
+        :type s: str
+        :rtype: int
+        """
+        tokens = tokenizeWithParens(s)
+        # print(tokens)
+        stack = []
+        for t in tokens:
+            if t.isdigit():
+                stack.append(int(t))
+            elif t in '(+-':
+                stack.append(t)
+            elif t == ')':
+                # print(stack)
+                subStack = []
+                while stack[-1] != '(':
+                    num = stack.pop()
+                    op = stack.pop()
+                    if op == '+':
+                        subStack.append(num)
+                    elif op == '-':
+                        subStack.append(-1 * num)
+                    elif op == '(':
+                        subStack.append(num)
+                        break
+                # print(sum(subStack))
+                stack.append(sum(subStack))
+        subStack = []
+        while stack:
+            num = stack.pop()
+            op = '+'
+            if stack:
+                op = stack.pop()
+            if op == '+':
+                subStack.append(num)
+            elif op == '-':
+                subStack.append(-1 * num)
+            elif op == '(':
```

Comment:
This case is not necessary -- you've already coalesced all the parentheticals into single values in your loop above. There will never be a '(' or ')' in the stack at this point.

Reply:
The case is necessary because a break is required to end the loop. I believe a better solution would be to instead replace the `while stack[-1] != '(':` with `while True:`